File: python/teach_SRP.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Single Responsibility Principle - SRP
# I do what I am responsible for
class Human(object):
    """
    class Human
    Incapsulate:
        - name
        - age
        - weight
    """
    def __init__(self, inName: str, inAge: int, inWeight: float):
        self.__name = inName
        self.__age = inAge
        self.__weight = inWeight

    def __str__(self):
        return f"Human: {self.__name} | {self.__age} | {self.__weight}"

    def go(self):
        print(f"{self.__name} GO!")

    @property
    def name(self):
        return self.__name

    @property
    def age(self):
        return self.__age

    @property
    def weight(self):
        return self.__weight


if __name__ == "__main__":
    human = Human("Lev Grodzickiy", 32, 111.3)

File: cpp/teach_SRP.cpp
#include <iostream>
#include <string>

// Single Responsibility Principle - SRP
// I do what I am responsible for

//
//  Human
//
class Human{
public:
    friend std::ostream& operator<< (std::ostream& out, const Human& h);

    explicit Human(std::string inName, int inAge, float inWeight):
        m_name(inName),
        m_age(inAge),
        m_weight(inWeight)
    {}

    void go(){
        std::cout << m_name << " GO!" << std::endl;
    }

    std::string getName() const{
        return m_name;
    }

    int getAge() const{
        return m_age;
    }

    float getWieght() const{
        return m_weight;
    }

private:
    std::string m_name;
    int         m_age;
    float       m_weight;

};

std::ostream& operator<< (std::ostream& out, const Human& h){
    out << "Human: " 
    << h.m_name << " | "
    << h.m_age << " | "
    << h.m_weight;

    return out;
}


int main(){
    Human human("Lev Grodzitskiy", 32, 111.3);
    std::cout << human << std::endl;

    std::cout << "Done!" << std::endl;
    return 0;
}
